### PR TITLE
OSDOCS-6109#Remove vSphere opt-in for migration

### PR DESCRIPTION
**Pull request by lpettyjo**
Version(s): 4.14+

Issue: https://issues.redhat.com/browse/OSDOCS-6109

Related PRs to change earlier OCP version for upgrade issue:
4.12: https://github.com/openshift/openshift-docs/pull/65546
4.13: https://github.com/openshift/openshift-docs/pull/65381

Link to docs preview: 
Migration topic: http://file.rdu.redhat.com/lpettyjo/OSDOCS-6109/storage/container_storage_interface/persistent-storage-csi-migration.html

vSphere CSI driver topic: https://60119--docspreview.netlify.app/openshift-enterprise/latest/storage/container_storage_interface/persistent-storage-csi-vsphere.html

QE review:
- [x] QE has approved this change.

Additional information:

**PTAL**: @gnufied, @gcharot, @ropatil010

### CODE REVIEW

**Comment by ropatil010** [comment]
LGTM
Thanks..!

File: modules/persistent-storage-csi-migration-vsphere.adoc [hunk 1]
```diff
@@ -25,62 +23,75 @@ Carefully review the following consequences before opting in to migration:
...
 
-=== Using the web console to opt in to automatic CSI migration
+Alternatively, if you do not want to upgrade vSphere, you can proceed with an {product-title} upgrade by performing an administrator acknowledgment:
 
-==== Prerequisites
+[source, cli]
+----
+oc -n openshift-config patch cm admin-acks --patch '{"data":{"ack-<version>-api-removals-in-4.14":"true"}}' --type=merge <1>
+----
+<1> Where `<version>` is the {product-title} and associated Kubernetes versions that you are upgrading from: either `4.13-kube-1.27` or `4.12-kube-1.26`.
```
Comment (gnufied) [inline modules/persistent-storage-csi-migration-vsphere.adoc:32]: The values are hardcoded and can't be replaced and hence there is no need of a note

File: modules/persistent-storage-csi-migration-vsphere.adoc [hunk 2]
```diff
@@ -25,62 +23,75 @@ Carefully review the following consequences before opting in to migration:
...
 
-* Access to the {product-title} web console.
+[IMPORTANT]
+====
+If you do *not* upgrade to vSphere 7.0 Update 3L or 8.0 Update 2 and use an administrator acknowledgment to upgrade to {product-title} 4.14, link:https://access.redhat.com/node/7011683[known issues can occur that can cause the migration to fail].
+====
```
Comment (gnufied) [inline modules/persistent-storage-csi-migration-vsphere.adoc:37]: > can cause the migration to fail

Migration will succeed regardless but we should word that - one or more pods that use in-tree vSphere volumes may fail to come up in certain cases and additional manual intervention will be necessary for bringing up these pods. Please refer to KCS for more details.

File: modules/persistent-storage-csi-migration-vsphere.adoc [hunk 3]
```diff
@@ -25,62 +23,75 @@ Carefully review the following consequences before opting in to migration:
...
-* Access to the cluster with cluster-admin privileges.
+//=== Using the web console to opt in to automatic CSI migration
+//
+//==== Prerequisites
 
-===== Procedure
+//* Access to the {product-title} web console.
 
-To opt in to automatic CSI migration for vSphere:
+//* Access to the cluster with cluster-admin privileges.
```
Comment (gnufied) [inline modules/persistent-storage-csi-migration-vsphere.adoc:45]: Are we commenting on these items? We probably will have to remove them altogether.

File: modules/persistent-storage-csi-migration-vsphere.adoc [hunk 4]
```diff
@@ -25,62 +23,75 @@ Carefully review the following consequences before opting in to migration:
...
+
+//[source, cli]
+//----
+//oc patch storage cluster --type=merge -p '{"spec":{"vsphereStorageDriver":"CSIWithMigrationDriver"}}'
+//----
+
+//[NOTE]
+//====
+//You can determine when migration is complete by verifying that the `VSphereMigrationControllerAvailable` condition is set to "true" in the `Storage` object.
+//====
```
Comment (gnufied) [inline modules/persistent-storage-csi-migration-vsphere.adoc:97]: All of this needs to be removed.

File: modules/persistent-storage-csi-migration-vsphere.adoc [hunk 5]
```diff
@@ -7,16 +7,14 @@
 = vSphere automatic migration
 
 == New installations of {product-title}
-For new installations of {product-title} 4.13, or later, automatic migration is enabled by default.
+For new installations of {product-title} 4.14, or later, automatic migration is enabled by default.
 
 == Upgrading {product-title}
-When upgrading from {product-title} 4.12, or earlier, to 4.13, automatic CSI migration for vSphere only occurs if you opt in.
+If you have vSphere in-tree persistent volumes (PVs) and want to upgrade from {product-title} 4.12 or 4.13 to 4.14, then you must upgrade to vSphere 7.0 Update 3L or 8.0 Update 2, or later. After upgrading vSphere, your {product-title} upgrade can occur and automatic migration is enabled by default.
```
Comment (gnufied) [inline modules/persistent-storage-csi-migration-vsphere.adoc:13]: >  vSphere in-tree persistent volumes (PVs)

This is not completely accurate and should be reworded on the lines of - "If you are using in-tree vSphere volumes".

File: modules/persistent-storage-csi-migration-vsphere.adoc [hunk 6]
```diff
@@ -7,16 +7,14 @@
...
 
 [IMPORTANT]
 ====
-Carefully review the following consequences before opting in to migration:
-
-* link:https://access.redhat.com/node/7011683[There are known issues that can occur that can cause the migration to fail].
+Migration includes significant consequences:
```
Comment (gnufied) [inline modules/persistent-storage-csi-migration-vsphere.adoc:17]: This note looks pretty scary. Where did we get wording for: 

```
Migration is a disruptive process. The Machine Config Operator (MCO) has to update kubelet, which means degrading each node as it rolls out the new MachineConfig to each machine.
```

?

File: modules/persistent-storage-csi-migration-vsphere.adoc [hunk 7]
```diff
@@ -7,16 +7,14 @@
 = vSphere automatic migration
 
 == New installations of {product-title}
-For new installations of {product-title} 4.13, or later, automatic migration is enabled by default.
+For new installations of {product-title} 4.14, or later, automatic migration is enabled by default.
```
[the remark this answers is not in the document]
Comment (gcharot) [reply modules/persistent-storage-csi-migration-vsphere.adoc:10]: yes since we explicitly mention "new installation" 4.13 is correct

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 5 above)
Comment (gcharot) [reply modules/persistent-storage-csi-migration-vsphere.adoc:13]: > > vSphere in-tree persistent volumes (PVs)
> 
> This is not completely accurate and should be reworded on the lines of - "If you are using in-tree vSphere volumes".

+1 and if CSI migration is NOT already enabled



> I would also avoid using the word `MUST` when we provide a workaround right below. We should say something like `it is recommended to upgrade`. Also lets be explicit and mention that both vCenter and ESXI host should be upgraded.

Agree, it's not a must since there is a way to bypass it. It is strongly recommended and upgrades will be blocked unless the vsphere version are running the correct version or an admin provides the admin ack.
+1 on mentioning vcenter and ESXi hosts

File: modules/persistent-storage-csi-migration-vsphere.adoc [hunk 8]
```diff
@@ -25,62 +23,75 @@ Carefully review the following consequences before opting in to migration:
 * Migration is a disruptive process. The Machine Config Operator (MCO) has to update kubelet, which means degrading each node as it rolls out the new MachineConfig to each machine.
 ====
 
-=== Using the web console to opt in to automatic CSI migration
+Alternatively, if you do not want to upgrade vSphere, you can proceed with an {product-title} upgrade by performing an administrator acknowledgment:
```
[the remark this answers is not in the document]
Comment (gcharot) [reply modules/persistent-storage-csi-migration-vsphere.adoc:26]: technically it's an update not an upgrade, we are not asking customers to upgrade to vsphere 8

File: modules/persistent-storage-csi-migration-vsphere.adoc [hunk 9]
```diff
@@ -25,62 +23,75 @@ Carefully review the following consequences before opting in to migration:
...
 
-==== Prerequisites
+[source, cli]
+----
+oc -n openshift-config patch cm admin-acks --patch '{"data":{"ack-<version>-api-removals-in-4.14":"true"}}' --type=merge <1>
+----
```
[the remark this answers is not in the document]
Comment (gcharot) [reply modules/persistent-storage-csi-migration-vsphere.adoc:31]: +1

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 2 above)
Comment (gcharot) [reply modules/persistent-storage-csi-migration-vsphere.adoc:37]: +1 problems are with pods that are using these in-tree volumes. I wouldn't care too much about details here, everything is in KCS.
Knows issues can occur, please carefully read the knowledge base article before proceeding with the administrator acknowledgement.

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 4 above)
Comment (gcharot) [reply modules/persistent-storage-csi-migration-vsphere.adoc:97]: yes all the opt-in 4.13 content needs to be removed

File: storage/container_storage_interface/persistent-storage-csi-vsphere.adoc [hunk 10]
```diff
@@ -24,15 +24,7 @@ To create CSI-provisioned persistent volumes (PVs) that mount to vSphere storage
```
Comment (gcharot) [inline storage/container_storage_interface/persistent-storage-csi-vsphere.adoc:24]: need to update the OCP version and the CSI driver version above (3.0.2)
cc @gnufied

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 6 above)
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-vsphere.adoc:17]: Docs have been published: 
4.13: https://access.redhat.com/documentation/en-us/openshift_container_platform/4.13/html/storage/using-container-storage-interface-csi#persistent-storage-csi-google-cloud-file
4.12: https://access.redhat.com/documentation/en-us/openshift_container_platform/4.12/html/storage/using-container-storage-interface-csi#persistent-storage-csi-google-cloud-file

Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-vsphere.adoc:17]: @gnufied @gcharot Not sure who provided this wording. Gregory or Eric maybe? 

Do you have a recommendation on how you would like it reworded? I would have to also go back and change it in the 4.13 docs, if so.

File: storage/container_storage_interface/persistent-storage-csi-vsphere.adoc (hunk 10 above)
Comment (lpettyjo) [reply storage/container_storage_interface/persistent-storage-csi-vsphere.adoc:24]: Done!

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 6 above)
Comment (gcharot) [reply modules/persistent-storage-csi-migration-vsphere.adoc:17]: I don't remember adding this.
Regardless, since we don't have it for other drivers, I think we should remove it unless @gnufied thinks we need to highlight something specific that does not apply to vsphere.

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 1 above)
Comment (gcharot) [reply modules/persistent-storage-csi-migration-vsphere.adoc:32]: Right but since it's an OCP 4.14 doc page, maybe we should cover both upgrade scenarios?
If you're upgrading from 4.12
`4.12 command`
If you're upgrading from 4.13
`4.13 command`

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 6 above)
Comment (gnufied) [reply modules/persistent-storage-csi-migration-vsphere.adoc:17]: Yeah we should remove it altogether in 4.14

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 1 above)
Comment (gnufied) [reply modules/persistent-storage-csi-migration-vsphere.adoc:32]: yeah I agree. I think it might be worth listing the keys and commands separately though, rather than just templating them because templating requires customers to be smart about replacing the values. 

I would prefer to have something that customer can simply copy and paste in terminal. Lets list 4.12 and 4.13 acks separately.

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 6 above)
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-vsphere.adoc:17]: @gnufied Should I remove this same bullet point from 4.13? I'm updating that section of the docs anyway on a related PR to cover the upgrade 4.13->4.14 issues.

Comment (gnufied) [reply modules/persistent-storage-csi-migration-vsphere.adoc:17]: Yes it can be removed from 4.13. But not from 4.12 I guess where migration is still TP for vSphere.

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 7 above)
[the remark this answers is not in the document]
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-vsphere.adoc]: Done.

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 9 above)
[the remark this answers is not in the document]
Comment (gnufied) [reply modules/persistent-storage-csi-migration-vsphere.adoc:51]: This key is still wrong.

**Comment by gnufied** [comment]
We also need something for EUS to EUS page of 4.14.

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 4 above)
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-vsphere.adoc]: Done.

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 5 above)
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-vsphere.adoc]: Fixed.

File: modules/persistent-storage-csi-migration-vsphere.adoc [hunk 11]
```diff
@@ -6,82 +6,53 @@
...
+If you do *not* update to vSphere 7.0 Update 3L or 8.0 Update 2 and use an administrator acknowledgment to update to {product-title} 4.14, known issues can occur. link:https://access.redhat.com/node/7011683[Before proceeding with the administrator acknowledgement, carefully read this knowledge base article].
+====
 
-* Access to the cluster with cluster-admin privileges.
+[id="updating-openshift_from_4.12_{context}"]
+== Updating from {product-title} 4.12 to 4.14
+If you are using vSphere in-tree persistent volumes (PVs) and want to update from {product-title} 4.12 to 4.14, update vSphere vCenter and ESXI host to 7.0 Update 3L or 8.0 Update 2, otherwise the {product-title} update is blocked. After updating vSphere, your {product-title} update can occur and automatic migration is enabled by default.
 
-==== Procedure
+Alternatively, if you do not want to update vSphere, you can proceed with an {product-title} update by performing an administrator acknowledgment:
```
Comment (gcharot) [inline modules/persistent-storage-csi-migration-vsphere.adoc:41]: since there are two admin-ack to perform, maybe rephrase the sentence to make it clear that the two acks are required

File: modules/persistent-storage-csi-migration-vsphere.adoc [hunk 12]
```diff
@@ -6,82 +6,53 @@
...
-. On the *CustomResourceDefinitions* page, type "Storage" to find the `Storage` custom resource (CR).
-
-. Click the *Storage* CR.
+== New installations of {product-title}
+For new installations of {product-title} 4.13, or later, automatic migration is enabled by default.
 
-. On the *storages.operator.openshift.io* page, click the *Instances* tab.
+[id="updating-openshift_from_4.13_{context}"]
+== Updating from {product-title} 4.13 to 4.14
+If you are using vSphere in-tree persistent volumes (PVs) and want to update from {product-title} 4.13 to 4.14, update vSphere vCenter and ESXI host to 7.0 Update 3L or 8.0 Update 2, otherwise the {product-title} update is blocked. After updating vSphere, your {product-title} update can occur and automatic migration is enabled by default.
```
Comment (gcharot) [inline modules/persistent-storage-csi-migration-vsphere.adoc:23]: Unless CSI migration is already enabled.
cc @gnufied

**Comment by lpettyjo** [comment]
> We also need something for EUS to EUS page of 4.14.

@gnufied @gcharot talked to Updates writer who spoke to his devs and concluded that they did not need to update the EUS-to-EUS content.

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 12 above)
Comment (gnufied) [reply modules/persistent-storage-csi-migration-vsphere.adoc:23]: Yeah - so something on the lines of "If you are using vSphere in-tree PVs and have not opted-in to CSI migration in 4.13, update from 4.13 to 4.14 will be blocked......."

File: modules/persistent-storage-csi-migration-vsphere.adoc [hunk 13]
```diff
@@ -6,82 +6,53 @@
...
-* link:https://access.redhat.com/node/7011683[There are known issues that can occur that can cause the migration to fail].
+Migration includes significant consequences:
 
 * Enabling migration cannot be undone.
 
-* Migration can take awhile to complete depending on how many nodes are on the cluster.
-
-* Migration is a disruptive process. The Machine Config Operator (MCO) has to update kubelet, which means degrading each node as it rolls out the new MachineConfig to each machine.
+* Migration can take a while to complete depending on how many nodes are on the cluster.
 ====
```
Comment (gnufied) [inline modules/persistent-storage-csi-migration-vsphere.adoc:16]: This need not be there anymore in 4.14 docs. Migration is automatic and not something that can be opted-out.

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 9 above)
[the remark this answers is not in the document]
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-vsphere.adoc:51]: Fixed.

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 1 above)
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-vsphere.adoc]: Fixed.

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 3 above)
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-vsphere.adoc]: Outdated. Fixed.

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 4 above)
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-vsphere.adoc]: Outdated. Fixed.

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 11 above)
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-vsphere.adoc:41]: Done.

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 13 above)
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-vsphere.adoc:16]: @gnufied I know that they can't opt out, but does that mean that they don't need to know about these consequences?

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 12 above)
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-vsphere.adoc:23]: @gnufied @gcharot agreed not add more complexity to this statement in the

**Comment by lpettyjo** [comment]
FYI. Here are the PRs for the EUS-to-EUS PRs:
Here are the PRs:
[4.13](https://github.com/openshift/openshift-docs/pull/66676)
[4.14](https://github.com/openshift/openshift-docs/pull/66675)

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 13 above)
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-vsphere.adoc:16]: OK. Thanks. It's removed.

**Comment by gnufied** [comment]
lgtm

**Comment by gcharot** [comment]
LGTM

File: modules/persistent-storage-csi-migration-vsphere.adoc [hunk 14]
```diff
@@ -9,79 +9,41 @@
...
-
-==== Procedure
-
-To opt in to automatic CSI migration for vSphere, run the following command:
-
-[source, cli]
+[source, terminal]
 ----
-oc patch storage cluster --type=merge -p '{"spec":{"vsphereStorageDriver":"CSIWithMigrationDriver"}}'
+oc -n openshift-config patch cm admin-acks --patch '{"data":{"ack-4.13-kube-127-vsphere-migration-in-4.14":"true"}}' –type=merge
```
Comment (duanwei33) [inline modules/persistent-storage-csi-migration-vsphere.adoc:41]: @lpettyjo 
```suggestion
oc -n openshift-config patch cm admin-acks --patch '{"data":{"ack-4.13-kube-127-vsphere-migration-in-4.14":"true"}}' --type=merge
```

**Comment by duanwei33** [comment]
Only one typo, others looks good to me.
The scenario has been verified by QE.

File: modules/persistent-storage-csi-migration-vsphere.adoc (hunk 14 above)
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-vsphere.adoc:41]: Thanks! Fixed.
